internal/iapl/policy.go

```diff
@@ -257,20 +399,205 @@ func (v *policy) expandActionBindings() {
 	}
 }
 
+// expandRole creates a list of all permissions, and a resource containing
+// a list of relationship to all permissions.
+func (v *policy) expandRole() {
+	// 1. create a list of all permissions
+	perms := make(map[string]struct{})
+
+	for _, action := range v.ac {
+		perms[action.Name] = struct{}{}
+	}
+
+	for perm := range perms {
+		v.permissions = append(v.permissions, perm)
+	}
+
+	// 2. create a relationship for role owners
+	roleOwners := Relationship{
+		Relation:    RoleOwnerRelation,
+		TargetTypes: make([]types.TargetType, len(v.p.RBAC.RoleOwners)),
+	}
+
+	for i, owner := range v.p.RBAC.RoleOwners {
+		roleOwners.TargetTypes[i] = types.TargetType{Name: owner}
+	}
+
+	// 3. create a list of relationships for all permissions
+	permsRel := make([]Relationship, len(perms))
+
+	for i, perm := range v.permissions {
+		targettypes := make([]types.TargetType, len(v.p.RBAC.RoleRelationshipSubjects))
+
+		for j, subject := range v.p.RBAC.RoleRelationshipSubjects {
+			targettypes[j] = types.TargetType{Name: subject, SubjectIdentifier: "*"}
+		}
+
+		permsRel[i] = Relationship{
+			Relation:    perm + "_rel",
+			TargetTypes: targettypes,
+		}
+	}
+
+	// 4. create a role containing all the relationships shown above
+	var role ResourceType
+
+	permsRel = append(permsRel, roleOwners)
+
+	if _, ok := v.rt[v.p.RBAC.RoleResource]; ok {
+		role = v.rt[v.p.RBAC.RoleResource]
+		role.Relationships = permsRel
+	} else {
+		role = ResourceType{
+			Name:          v.p.RBAC.RoleResource,
+			Relationships: permsRel,
+		}
+	}
+
+	v.rt[role.Name] = role
+
+	// 5. create a list of role owners
+	for _, owner := range v.p.RBAC.RoleOwners {
+		v.roleOwners[owner] = struct{}{}
+	}
+}
+
+func (v *policy) expandRoleBinding() {
+	// 1. create relationship to role
+	role := Relationship{
+		Relation: RolebindingRoleRelation,
+		TargetTypes: []types.TargetType{
+			{Name: v.p.RBAC.RoleResource},
+		},
+	}
+
+	// 2. create relationship to subjects
+	subjects := Relationship{
+		Relation:    RolebindingSubjectRelation,
+		TargetTypes: v.p.RBAC.RoleBindingSubjects,
+	}
+
+	// 3. create a list of action-bindings representing permissions for all the
+	// actions
+	actionbindings := make([]ActionBinding, len(v.ac))
+	i := 0
```

Comment:
What's the intent of this? `make([]ActionBinding, 0, len(v.ac))` and then `actionBindings = append(actionBindings, ab)` would accomplish the same thing, if a single allocation is what you're looking for.